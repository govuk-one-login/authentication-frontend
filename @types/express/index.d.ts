import type pino from "pino";
import type { Taxonomy } from "../utils/taxonomy.ts";

declare global {
  namespace Express {
    export interface Request {
      i18n?: {
        language?: string;
      };
      t: TFunction;
      csrfToken?: () => string;
      log: pino.Logger;
      generatedSessionId?: string;
    }

    export interface Locals {
      // Session IDs
      sessionId?: string;
      clientSessionId?: string;
      persistentSessionId?: string;
      clientId?: string;

      // Security
      scriptNonce?: string;
      csrfToken?: string;

      // Analytics
      ga4ContainerId?: string;
      isGa4Enabled?: string;
      analyticsCookieDomain?: string;
      analyticsProperties?: {
        httpStatusCode: number;
        contentId: string;
        taxonomy: Taxonomy;
      };

      // Language
      languageToggleEnabled?: boolean;
      language?: string;
      htmlLang?: string;

      // App
      strategicAppChannel?: boolean;
      webChannel?: boolean;
      genericAppChannel?: boolean;
      isApp?: boolean;

      // Feature flags
      supportPasskeyUsage?: boolean;
      supportPasskeyRegistration?: boolean;

      // Misc
      currentUrl?: URL;
      showTestBanner?: boolean;
      accountManagementUrl?: string;
      contactUsLinkUrl?: string;
    }
  }
}
