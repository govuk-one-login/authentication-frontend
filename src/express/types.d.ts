import 'express-serve-static-core';

declare global {
  namespace Express {
    interface Locals {
      sessionId: string;
      strategicAppChannel: boolean;
      webChannel: boolean;
      genericAppChannel: boolean;
      isApp: boolean
    }
    interface Response {
      locals: Locals;
    }
  }
}
