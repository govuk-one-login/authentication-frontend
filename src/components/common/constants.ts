import { PATH_NAMES } from "../../app.constants";
import { getNextState } from "./state-machine/state-machine";

export const SECURITY_CODE_ERROR = "actionType";

export enum SecurityCodeErrorType {
  MfaMaxCodesSent = "mfaMaxCodesSent",
  MfaBlocked = "mfaBlocked",
  MfaMaxRetries = "mfaMaxRetries",
  OtpMaxCodesSent = "otpMaxCodesSent",
  OtpBlocked = "otpBlocked",
  OtpMaxRetries = "otpMaxRetries",
  EmailMaxCodesSent = "emailMaxCodesSent",
  EmailBlocked = "emailBlocked",
  EmailMaxRetries = "emailMaxRetries",
  AuthAppMfaNoCodeValidator = "authAppMfaNoCodeValidator",
  AuthAppMfaMaxRetries = "authAppMfaMaxRetries",
  ChangeSecurityCodesEmailMaxCodesSent = "changeSecurityCodesEmailMaxCodesSent",
  ChangeSecurityCodesEmailBlocked = "changeSecurityCodesEmailBlocked",
  ChangeSecurityCodesEmailMaxRetries = "changeSecurityCodesEmailMaxRetries",
  InvalidPasswordResetCodeMaxRetries = "invalidPasswordResetCodeMaxRetries",
}

export const ERROR_CODES = {
  SESSION_ID_MISSING_OR_INVALID: 1000,
  RESET_PASSWORD_LINK_MAX_RETRIES_REACHED: 1022,
  RESET_PASSWORD_LINK_BLOCKED: 1023,
  NEW_PASSWORD_SAME_AS_EXISTING: 1024,
  ENTERED_INVALID_MFA_MAX_TIMES: 1027,
  VERIFY_EMAIL_CODE_REQUEST_BLOCKED: 1029,
  VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED: 1030,
  VERIFY_PHONE_NUMBER_MAX_CODES_SENT: 1032,
  VERIFY_EMAIL_MAX_CODES_SENT: 1031,
  ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES: 1033,
  ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES: 1034,
  MFA_SMS_MAX_CODES_SENT: 1025,
  MFA_CODE_REQUESTS_BLOCKED: 1026,
  INVALID_MFA_CODE: 1035,
  INVALID_VERIFY_EMAIL_CODE: 1036,
  INVALID_VERIFY_PHONE_NUMBER_CODE: 1037,
  ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES: 1039,
  INVALID_PASSWORD_MAX_ATTEMPTS_REACHED: 1028,
  RESET_PASSWORD_INVALID_CODE: 1021,
  PASSWORD_IS_COMMON: 1040,
  AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED: 1042,
  AUTH_APP_INVALID_CODE: 1043,
  ACCOUNT_LOCKED: 1045,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT: 1046,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED: 1047,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE: 1048,
};

export const ERROR_CODE_MAPPING: { [p: string]: string } = {
  [ERROR_CODES.ACCOUNT_LOCKED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.MFA_SMS_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxCodesSent
  ),
  [ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpMaxCodesSent
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.OtpMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxRetries
  ),
  [ERROR_CODES.ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.InvalidPasswordResetCodeMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxCodesSent
  ),
  [ERROR_CODES.AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.AuthAppMfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxCodesSent
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxRetries
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_WAIT"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailBlocked
    ),
};

export function getErrorPathByCode(errorCode: number): string | undefined {
  const nextPath = ERROR_CODE_MAPPING[errorCode.toString()];

  if (!nextPath) {
    return undefined;
  }

  return nextPath;
}

export function pathWithQueryParam(
  path: string,
  queryParam?: string,
  value?: string | SecurityCodeErrorType
): string {
  if (queryParam && value) {
    const queryParams = new URLSearchParams({
      [queryParam]: value,
    }).toString();

    return path + "?" + queryParams;
  }

  return path;
}

export function getNextPathAndUpdateJourney(
  req: any,
  path: string,
  event: string,
  ctx?: any,
  sessionId?: string
): string {
  const nextState = getNextState(path, event, ctx);

  req.session.user.journey = {
    nextPath: nextState.value,
    optionalPaths:
      Object.keys(nextState.meta).length > 0
        ? nextState.meta["AUTH." + nextState.value].optionalPaths
        : [],
  };

  req.log.info(
    `User journey transitioned from ${req.path} to ${nextState.value} with session id ${sessionId}`
  );

  if (!nextState) {
    throw Error(
      `Invalid user journey. No transition found from ${path} with event ${event}`
    );
  }

  return nextState.value;
}

export const JOURNEY_TYPE = {
  SIGN_IN: "sign-in",
  CREATE_ACCOUNT: "create-account",
};
