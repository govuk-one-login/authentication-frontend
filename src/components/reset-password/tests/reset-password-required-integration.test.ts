import { describe } from "mocha";
import { expect, request, sinon } from "../../../../test/utils/test-utils.js";
import nock from "nock";
import * as cheerio from "cheerio";
import { MFA_METHOD_TYPE, PATH_NAMES } from "../../../app.constants.js";
import decache from "decache";
import {
  noInterventions,
  setupAccountInterventionsResponse,
} from "../../../../test/helpers/account-interventions-helpers.js";
import { NextFunction, Request, Response } from "express";
import { getPermittedJourneyForPath } from "../../../../test/helpers/session-helper.js";
import { buildMfaMethods } from "../../../../test/helpers/mfa-helper.js";

describe("Integration::reset password required", () => {
  let token: string | string[];
  let cookies: string;
  let app: any;
  let baseApi: string;

  const ENDPOINT = PATH_NAMES.RESET_PASSWORD_REQUIRED;

  before(async () => {
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    const sessionMiddleware = await import(
      "../../../middleware/session-middleware"
    );

    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (
        req: Request,
        res: Response,
        next: NextFunction
      ): void {
        res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";
        req.session.user = {
          email: "test@test.com",
          mfaMethods: buildMfaMethods({ phoneNumber: "7867" }),
          journey: getPermittedJourneyForPath(ENDPOINT),
          isAuthenticated: true,
          isAccountPartCreated: false,
          accountRecoveryVerifiedMfaType: MFA_METHOD_TYPE.SMS,
        };

        next();
      });
    app = await (await import("../../../app")).createApp();
    baseApi = process.env.FRONTEND_API_BASE_URL;
    setupAccountInterventionsResponse(baseApi, noInterventions);

    await request(app, (test) => test.get(ENDPOINT), {
      expectAnalyticsPropertiesMatchSnapshot: false,
    }).then((res) => {
      const $ = cheerio.load(res.text);
      token = $("[name=_csrf]").val();
      cookies = res.headers["set-cookie"];
    });
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  after(() => {
    sinon.restore();
    app = undefined;
  });

  it("should return reset password page", async () => {
    setupAccountInterventionsResponse(baseApi, noInterventions);

    await request(app, (test) => test.get(ENDPOINT).expect(200));
  });

  it("should return error when csrf not present", async () => {
    await request(app, (test) =>
      test
        .post(ENDPOINT)
        .type("form")
        .send({
          password: "password",
        })
        .expect(403)
    );
  });

  it("should return validation error when password not entered", async () => {
    await request(app, (test) =>
      test
        .post(ENDPOINT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          password: "",
          "confirm-password": "",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#password-error").text()).to.contains(
            "Enter your password"
          );
        })
        .expect(400)
    );
  });

  it("should return validation error when passwords don't match", async () => {
    await request(app, (test) =>
      test
        .post(ENDPOINT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          password: "sadsadasd33da",
          "confirm-password": "sdnnsad99d",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#confirm-password-error").text()).to.contains(
            "Enter the same password in both fields"
          );
        })
        .expect(400)
    );
  });

  it("should return validation error when password less than 8 characters", async () => {
    await request(app, (test) =>
      test
        .post(ENDPOINT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          password: "dad",
          "confirm-password": "",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#password-error").text()).to.contains(
            "Your password must be at least 8 characters long and must include letters and numbers"
          );
        })
        .expect(400)
    );
  });

  it("should return validation error when password is amongst most common passwords", async () => {
    nock(baseApi).post("/reset-password").once().reply(400, { code: 1040 });

    await request(app, (test) =>
      test
        .post(ENDPOINT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          password: "password123",
          "confirm-password": "password123",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#password-error").text()).to.contains(
            "Enter a stronger password. Do not use very common passwords, such as ‘password’ or a sequence of numbers."
          );
        })
        .expect(400)
    );
  });

  it("should return error when new password is the same as existing password", async () => {
    nock(baseApi).post("/reset-password").once().reply(400, { code: 1024 });

    await request(app, (test) =>
      test
        .post(ENDPOINT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          password: "p@ssw0rd-123",
          "confirm-password": "p@ssw0rd-123",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#password-error").text()).to.contains(
            "You are already using that password. Enter a different password"
          );
        })
        .expect(400)
    );
  });

  it("should return validation error when no numbers present in password", async () => {
    await request(app, (test) =>
      test
        .post(ENDPOINT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          password: "testpassword",
          "confirm-password": "testpassword",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#password-error").text()).to.contains(
            "Your password must be at least 8 characters long and must include letters and numbers"
          );
        })
        .expect(400)
    );
  });

  it("should return validation error when password all numeric", async () => {
    await request(app, (test) =>
      test
        .post(ENDPOINT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          password: "222222222222222",
          "confirm-password": "222222222222222",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#password-error").text()).to.contains(
            "Your password must be at least 8 characters long and must include letters and numbers"
          );
        })
        .expect(400)
    );
  });

  it("should redirect to /auth-code when valid password entered", async () => {
    nock(baseApi).post("/reset-password").once().reply(204);
    nock(baseApi).post("/login").once().reply(200);
    nock(baseApi).post("/mfa").once().reply(204);

    await request(app, (test) =>
      test
        .post(ENDPOINT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          password: "Testpassword1",
          "confirm-password": "Testpassword1",
        })
        .expect("Location", PATH_NAMES.AUTH_CODE)
        .expect(302)
    );
  });
});
