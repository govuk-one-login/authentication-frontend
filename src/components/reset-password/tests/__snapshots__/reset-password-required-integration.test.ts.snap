// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration::reset password required should redirect to /auth-code when valid password entered 1`] = `
Object {
  "contentId": "95e26313-bc2f-49bc-bc62-fd715476c1d9",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::reset password required should return error when csrf not present 1`] = `
Object {
  "contentId": undefined,
  "taxonomyLevel1": undefined,
  "taxonomyLevel2": undefined,
  "taxonomyLevel3": undefined,
  "taxonomyLevel4": undefined,
  "taxonomyLevel5": undefined,
}
`;

exports[`Integration::reset password required should return error when new password is the same as existing password 1`] = `
Object {
  "contentId": "c8520c6c-9f09-4edf-8c99-7123a3991cfc",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "account recovery",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::reset password required should return reset password page 1`] = `
Object {
  "contentId": "c8520c6c-9f09-4edf-8c99-7123a3991cfc",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "account recovery",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::reset password required should return validation error when no numbers present in password 1`] = `
Object {
  "contentId": "c8520c6c-9f09-4edf-8c99-7123a3991cfc",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "account recovery",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::reset password required should return validation error when password all numeric 1`] = `
Object {
  "contentId": "c8520c6c-9f09-4edf-8c99-7123a3991cfc",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "account recovery",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::reset password required should return validation error when password is amongst most common passwords 1`] = `
Object {
  "contentId": "c8520c6c-9f09-4edf-8c99-7123a3991cfc",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "account recovery",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::reset password required should return validation error when password less than 8 characters 1`] = `
Object {
  "contentId": "c8520c6c-9f09-4edf-8c99-7123a3991cfc",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "account recovery",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::reset password required should return validation error when password not entered 1`] = `
Object {
  "contentId": "c8520c6c-9f09-4edf-8c99-7123a3991cfc",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "account recovery",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::reset password required should return validation error when passwords don't match 1`] = `
Object {
  "contentId": "c8520c6c-9f09-4edf-8c99-7123a3991cfc",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "account recovery",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;
